### PR TITLE
Add some initial minimal docs for each check

### PR DESCRIPTION
**Pull request by swissspidy**
Each _**check**_ should have a short description that can be displayed in the admin / CLI, and a link to a web page with a more detailed documentation, which can be more easily maintained.

**Note:** in the future we can add docs for each _error_ or _warning_ reported by a check. But that's out of scope for this PR.

In the admin it could look like this:

Screenshot 2024-07-29 at 14 51 31

Screenshot 2024-07-29 at 14 51 40

See #448
See #316

**To-do:**

- [ ] Rephrase short descriptions, update URLs
- [ ] Create handbook pages for some checks which don't have any matching documentation pages yet
- [ ] Maybe add a `get_name()` method so every check has a human-readable name
- [ ] Maybe move the check slug to a `get_slug()` method
- [x] Maybe centralize documentation somewhere, e.g. under https://developer.wordpress.org/plugins/
- [x] Actually display the documentation in the admin / CLI context

### CODE REVIEW

**Comment by swissspidy** [comment]
> Actually display the documentation in the admin / CLI context

Briefly looked into this and and this needs some work for sure.  The data from`Check_Result` is essentially what's being returned via admin-ajax, and `Check_Result` can _theoretically_ hold errors & warnings from multiple checks (Even though it's currently only 1 at the time). So we would need to somehow attach more information to the `Check_Result` (check name, description, and URL) for each check and display that in the admin in the right place. Aside: there's also no connection between an added error and which check it's coming from.

Then there are also some detailed descriptions for specific errors, as shown in #316. Those are better put on the handbook website.

**Comment by davidperezgar** [comment]
Hello, you're right, every check has their own description. I'd prefer to have them all in a folder with html files. That's how we have it already in our tool. And it's much easier to maintain these files.

**Comment by swissspidy** [comment]
I still think it makes more sense to put the long versions on the handbook website for easier maintenance and reference.

If you want them as part of the plugin, then they all need to be translatable, so an HTML file is not good enough.

**Comment by davidperezgar** [comment]
Yes, you're right in make it translatable. It's not that long an is a content of years and years reviewing, so I think is very important to have it.

**Comment by davidperezgar** [comment]
I'm reviewing this PR but I see that description is for all checks, and We do need to have a description in every check. Maybe we could change the description function to have parameters, so we can pass a slug of check and it give the solution. 

I give an example of a description of check:

> ## Not permitted files
> 
> A plugin typically consists of files related to the plugin functionality (php, js, css, txt, md) and maybe some multimedia files (png, svg, jpg) and / or data files (json, xml).
> 
> We have detected files that are not among of the files normally found in a plugin, are they necessary? If not, then those won't be allowed.
> 
> Example(s) from your plugin:


What do you think?

**Comment by swissspidy** [comment]
In the admin it could look like this:

Screenshot 2024-07-29 at 14 51 31

Screenshot 2024-07-29 at 14 51 40

**Comment by davidperezgar** [comment]
Yes, that's good option, but I'd add it as well in the results of the checks.

**Comment by swissspidy** [comment]
> Yes, that's good option, but I'd add it as well in the results of the checks.

Let's do that in a follow-up issue.

As per https://github.com/WordPress/plugin-check/pull/461#issuecomment-2206239406 that is going to be more complicated. The results do not have any information about the check attached, and we don't currently send that information via the Ajax responses.

Also keep in mind that result ≠ check. 1 Check can have hundreds of results. So we'd need to figure out how to best display the information without adding too much noise.

File: includes/Admin/Admin_Page.php [hunk 1]
```diff
@@ -91,6 +94,80 @@ public function add_and_initialize_page() {
...
+			<dd>
+				<?php echo wp_kses( $check->get_description(), array( 'code' => array() ) ); ?>
+				<br>
+				<a href="<?php echo esc_url( $check->get_documentation_url() ); ?>">
+					<?php esc_html_e( 'Learn more', 'plugin-check' ); ?>
+				</a>
+			</dd>
+			<?php
+
+			echo '';
```
Comment (mukeshpanchal27) [inline includes/Admin/Admin_Page.php]:
```suggestion
```

File: includes/Checker/Checks/Performance/Enqueued_Resources_Check.php [hunk 2]
```diff
@@ -47,4 +47,30 @@ protected function get_args() {
...
+	 * Gets the description for the check.
+	 *
+	 * Every check must have a short description explaining what the check does.
+	 *
+	 * @since 1.1.0
+	 *
+	 * @return string Description.
+	 */
+	public function get_description(): string {
+		return __( 'Checks whether scripts and styles are properly enqueued using the recommended way', 'plugin-check' );
```
Comment (mukeshpanchal27) [inline includes/Checker/Checks/Performance/Enqueued_Resources_Check.php]:
```suggestion
		return __( 'Checks whether scripts and styles are properly enqueued using the recommended way.', 'plugin-check' );
```

File: includes/Checker/Checks/Performance/Non_Blocking_Scripts_Check.php [hunk 3]
```diff
@@ -240,4 +240,30 @@ private function get_viewable_post_types() {
...
+	 * Gets the description for the check.
+	 *
+	 * Every check must have a short description explaining what the check does.
+	 *
+	 * @since 1.1.0
+	 *
+	 * @return string Description.
+	 */
+	public function get_description(): string {
+		return __( 'Checks whether scripts and styles are enqueued using a recommended loading strategy', 'plugin-check' );
```
Comment (mukeshpanchal27) [inline includes/Checker/Checks/Performance/Non_Blocking_Scripts_Check.php]:
```suggestion
		return __( 'Checks whether scripts and styles are enqueued using a recommended loading strategy.', 'plugin-check' );
```

File: includes/Checker/Checks/Security/Direct_DB_Queries_Check.php [hunk 4]
```diff
@@ -47,4 +47,30 @@ protected function get_args() {
...
+	 * Gets the description for the check.
+	 *
+	 * Every check must have a short description explaining what the check does.
+	 *
+	 * @since 1.1.0
+	 *
+	 * @return string Description.
+	 */
+	public function get_description(): string {
+		return __( 'Checks the usage of direct database queries, which should be avoided', 'plugin-check' );
```
Comment (mukeshpanchal27) [inline includes/Checker/Checks/Security/Direct_DB_Queries_Check.php]:
```suggestion
		return __( 'Checks the usage of direct database queries, which should be avoided.', 'plugin-check' );
```

**Review by mukeshpanchal27 — COMMENTED** [review]
The changes looks good to me. Left some nit-pick

Could you please update https://github.com/WordPress/plugin-check/tree/trunk/docs, the static and runtime check?

**Comment by davidperezgar** [comment]
I propose this. Make the title of the result, description and link in the same page. 

Yes it's longer, but our experience reviewing plugins is that developers need help solving the issues.

screenshot-upload

**Comment by swissspidy** [comment]
@davidperezgar but that‘s the message for an individual error, not a whole check. Let‘s handle that in a follow-up issue, as per my above comment

File: includes/Checker/Checks/Performance/Enqueued_Styles_Size_Check.php [hunk 5]
```diff
@@ -264,4 +264,34 @@ private function get_viewable_post_types() {
...
+	 *
+	 * Every check must have a short description explaining what the check does.
+	 *
+	 * @since 1.1.0
+	 *
+	 * @return string Description.
+	 */
+	public function get_description(): string {
+		return sprintf(
+		/* translators: %s: Script size threshold. */
```
Comment (westonruter) [inline includes/Checker/Checks/Performance/Enqueued_Styles_Size_Check.php]:
```suggestion
			/* translators: %s: Script size threshold. */
```

**Review by westonruter — APPROVED** [review]
One tiny nit, but pre-approving pending applying Mukesh's suggestions.

**Comment by davidperezgar** [comment]
Ok @swissspidy , make that follow up.

**Comment by davidperezgar** [comment]
We need for results, not whole check. The explanations for the results vary a lot between them.

**Review by ernilambar — APPROVED** [review]

**Review by mukeshpanchal27 — APPROVED** [review]
The documents changes can we done in followup PR.

**Pull request merged**